FSW/src/util/hkUtil.cpp

```diff
@@ -0,0 +1,184 @@
+    float digitalTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_DIGITAL_THERM);
+    
+    // update housekeeping data
+    latestHkSample.opticsTemp = voltageToTemp(opticsTempVoltage); 
+    latestHkSample.analogTemp = voltageToTemp(analogTempVoltage);
+    latestHkSample.digitalTemp = voltageToTemp(digitalTempVoltage);
+    latestHkSample.analogCurrent = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_AREG_CURR);
+    latestHkSample.digitalCurrent = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_DREG_CURR);   
+    latestHkSample.digitalRegPG = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_DREG_PG);
+    hkData[hkIndex] = latestHkSample; // update data array
+
+    // Check if values are in acceptable ranges
+    // TODO: Log a fault for every case
+    if (latestHkSample.opticsTemp > OPTICS_TEMP_MAX_SAFE) { } // optics temp too high
+    else if (latestHkSample.opticsTemp < OPTICS_TEMP_MIN_SAFE) { } // optics temp too low
+
+    if (latestHkSample.analogTemp > BOARD_TEMP_MAX_SAFE) { } // analog board temp too high
+    else if (latestHkSample.analogTemp < BOARD_TEMP_MIN_SAFE) { } // analog board temp too low
+
+    if (latestHkSample.digitalTemp > BOARD_TEMP_MAX_SAFE) { } // digital board temp too high
+    else if (latestHkSample.digitalTemp < BOARD_TEMP_MIN_SAFE) { } // digital board temp too low
+
+    if (latestHkSample.digitalRegPG > PG_VOLTAGE_MAX_EXPECTED) { } // digital regulator PG signal too high
+    else if (latestHkSample.digitalRegPG < PG_VOLTAGE_MIN_EXPECTED) { } // digital regulator PG signal too low
+
+}
+
+/* - - - - - - voltageToTemp - - - - - - *
+ * Usage:
+ *  Interpolates and returns the corresponding temperature for a given thermistor voltage reading
+ *  based on the data in thermLookup. 
+ *  Will extrapolate beyond max and min entries in thermLookup if needed
+ * 
+ * Inputs:
+ *  voltage - thermistor voltage
+ *  
+ * Outputs:
+ *  temperature - celsius, the temperature corresponding to thermistor voltage reading
+ */
+float voltageToTemp(float voltage)
+{
+    int thermLookupSize = sizeof(thermLookup) / sizeof(thermLookup[0]); // size of lookup table
+    
+    // find lower adjacent entry
+    int i = 0;
+    while (thermLookup[i].voltage > voltage && i < thermLookupSize - 1) { i++; }
+    
+    float tempGap = thermLookup[i - 1].temperature - thermLookup[i].temperature; // temp difference between upper and lower entries
+    float voltageGap = thermLookup[i - 1].voltage - thermLookup[i].voltage; // voltage difference between upper and lower entries
+    
+    // interpolate temperature
+    float temperature = tempGap*(voltage - thermLookup[i].voltage) / voltageGap + thermLookup[i].temperature;
+    return temperature;
+}
+
+/* - - - - - - timeSortHkData - - - - - - *
+ * Usage:
+ *  Sorts hkData in ascending cronological order and resets hkIndex to 0
+ *  call before using hkData in another function
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  none
+ */
+void timeSortHkData()
```

Comment:
How similar is this to the similar function that sorts the science data by ascending? Can they both be used for the same purpose, or does the `struct` nature of the HK data require this new function?

Reply:
Very similar. I had the same thought, but C++ really likes to know exactly what kind of data it is working with, so making generalized timeSortArray function would be more complicated and would likely use more memory than just having the two functions. 
This works for now, but in the future I am thinking of maybe replacing each array with a std::vector. They use more memory, but they are dynamically scalable, and have all kinds of awesome functions like push_back() and erase(), which would allow us to keep the vector sorted all the time.